src/cookies.js

```diff
@@ -83,27 +83,60 @@ function tryGetDocumentCookieNoInline(win) {
  *     - highestAvailableDomain: If true, set the cookie at the widest domain
  *       scope allowed by the browser. E.g. on example.com if we are currently
  *       on www.example.com.
- *     - domain: Explicit domain to set.
+ *     - domain: Explicit domain to set. domain overrides HigestAvailableDomain
  *     - allowOnProxyOrigin: Allow setting a cookie on the AMP Cache.
  */
 export function setCookie(win, name, value, expirationTime, opt_options) {
   checkOriginForSettingCookie(win, opt_options, name);
-  if (opt_options && opt_options.highestAvailableDomain) {
+  let domain = undefined;
+  // Respect explicitily set domain over higestAvailabeDomain
+  if (opt_options && opt_options.domain) {
+    domain = opt_options.domain;
+  } else if (opt_options && opt_options.highestAvailableDomain) {
+    domain = getHighestAvailableDomain(win);
+    if (!domain) {
+      return;
+    }
+  }
+  trySetCookie(win, name, value, expirationTime, domain);
+}
+
+/**
+ * Attemp to find the HighestAvailableDomain on
+ * @param {!Window} win
+ */
+export function getHighestAvailableDomain(win) {
+  // TODO(zhouyx@): Once we decide to allow publisher to put
+  // <meta name='amp-cookie-scope'>. Need to respect the meta first.
+  if (!isProxyOrigin(win.location.href)) {
+    // Use the set cookie hack to find the higestAvailableDomain on non proxy
+    // origin.
+
     const parts = win.location.hostname.split('.');
     let domain = parts[parts.length - 1];
+    const randomName = Math.random().toString();
     for (let i = parts.length - 2; i >= 0; i--) {
       domain = parts[i] + '.' + domain;
-      trySetCookie(win, name, value, expirationTime, domain);
-      if (getCookie(win, name) == value) {
-        return;
+      // Try set a cookie for testing only, expire after 1 sec
+      trySetCookie(win, randomName, 'delete', Date.now() + 1000, domain);
```

Comment:
I'm a bit unsure about what to do with the cookie name. Let me know what do you think.

Reply:
let's fix a name to something very special, for example: `-amp-cookie-test-tmp`.
to avoid a conflict, we can read all existing cookies and check. append counter to the name if a conflict is found

Reply:
sounds good. will do

Reply:
done